File: sdks/go/schema_evolution.go
package vectorizer

import (
	"net/url"
)

// RenameCollection atomically renames a collection.
//
// Calls POST /collections/{name}/rename with body {"new_name": newName}.
// The server retains the old name as an in-memory alias for one minor
// version so existing clients continue to work without reconfiguration.
// The alias does not survive a restart.
func (c *Client) RenameCollection(oldName, newName string) error {
	path := "/collections/" + url.PathEscape(oldName) + "/rename"
	body := map[string]string{"new_name": newName}
	return c.request("POST", path, body, nil)
}

// ReindexCollection rebuilds the HNSW index for a collection with new
// parameters.
//
// Calls POST /collections/{name}/reindex with body
// {"m": ..., "ef_construction": ..., "ef_search": ...}.
// No re-embedding is required — existing stored vectors are reused.
// Returns a ReindexJob with State == "completed" on success.
func (c *Client) ReindexCollection(name string, params *ReindexParams) (*ReindexJob, error) {
	path := "/collections/" + url.PathEscape(name) + "/reindex"
	var result ReindexJob
	if err := c.request("POST", path, params, &result); err != nil {
		return nil, err
	}
	return &result, nil
}

// SnapshotCollectionNative creates a native per-collection snapshot.
//
// Calls POST /collections/{name}/snapshot with an empty body.
// The server writes a gzip-compressed JSON snapshot and returns snapshot
// metadata including ID, collection name, creation time, and size.
func (c *Client) SnapshotCollectionNative(name string) (*NativeSnapshotInfo, error) {
	path := "/collections/" + url.PathEscape(name) + "/snapshot"
	var result NativeSnapshotInfo
	if err := c.request("POST", path, map[string]interface{}{}, &result); err != nil {
		return nil, err
	}
	return &result, nil
}

// ListCollectionSnapshotsNative lists all native snapshots for a collection.
//
// Calls GET /collections/{name}/snapshots.
// Returns snapshots newest-first as reported by the server.
func (c *Client) ListCollectionSnapshotsNative(name string) ([]NativeSnapshotInfo, error) {
	path := "/collections/" + url.PathEscape(name) + "/snapshots"
	var envelope struct {
		Snapshots []NativeSnapshotInfo `json:"snapshots"`
	}
	if err := c.request("GET", path, nil, &envelope); err != nil {
		return nil, err
	}
	return envelope.Snapshots, nil
}

// RestoreCollectionSnapshotNative restores a collection from a native snapshot.
//
// Calls POST /collections/{name}/snapshots/{id}/restore with an empty body.
// Drops the current in-memory state and replaces it with the snapshot data.
func (c *Client) RestoreCollectionSnapshotNative(name, snapshotID string) error {
	path := "/collections/" + url.PathEscape(name) + "/snapshots/" + url.PathEscape(snapshotID) + "/restore"
	return c.request("POST", path, map[string]interface{}{}, nil)
}

// ExplainSearch runs a vector search and returns the full HNSW execution trace.
//
// Calls POST /collections/{name}/explain with body {"vector": [...], "k": k}.
// The trace includes visited_nodes, ef_search, hnsw_search_ms,
// payload_filter_evals, quantization_score_ms, and total_ms.
// The results are identical to a normal search; the real code path is
// instrumented rather than a separate explain engine.
func (c *Client) ExplainSearch(collection string, vector []float32, k int) (*ExplainResponse, error) {
	path := "/collections/" + url.PathEscape(collection) + "/explain"
	body := map[string]interface{}{
		"vector": vector,
		"k":      k,
	}
	var result ExplainResponse
	if err := c.request("POST", path, body, &result); err != nil {
		return nil, err
	}
	return &result, nil
}

// ListSlowQueries returns entries from the slow-query ring buffer.
//
// Calls GET /slow_queries. Returns entries in the order they were recorded
// (oldest first). Use SetSlowQueryConfig to tune the threshold and capacity.
func (c *Client) ListSlowQueries() ([]SlowQueryEntry, error) {
	var envelope struct {
		Entries []SlowQueryEntry `json:"entries"`
	}
	if err := c.request("GET", "/slow_queries", nil, &envelope); err != nil {
		return nil, err
	}
	return envelope.Entries, nil
}

// SetSlowQueryConfig reconfigures the slow-query ring buffer.
//
// Calls POST /slow_queries/config with body
// {"threshold_ms": ..., "capacity": ...}.
// Existing entries are retained. If the new capacity is smaller than the
// current entry count the oldest entries are evicted by the server.
// Returns the updated configuration as echoed back by the server.
func (c *Client) SetSlowQueryConfig(config *SlowQueryConfig) (*SlowQueryConfig, error) {
	var result SlowQueryConfig
	if err := c.request("POST", "/slow_queries/config", config, &result); err != nil {
		return nil, err
	}
	return &result, nil
}

File: sdks/csharp/SchemaEvolution.cs
using Vectorizer.Models;

namespace Vectorizer;

public partial class VectorizerClient
{
    private sealed class SnapshotsEnvelope
    {
        [System.Text.Json.Serialization.JsonPropertyName("snapshots")]
        public List<NativeSnapshotInfo>? Snapshots { get; set; }
    }

    private sealed class SlowQueryEntriesEnvelope
    {
        [System.Text.Json.Serialization.JsonPropertyName("entries")]
        public List<SlowQueryEntry>? Entries { get; set; }
    }

    /// <summary>
    /// Atomically renames a collection (POST /collections/{name}/rename).
    /// The server retains the old name as an in-memory alias for one minor
    /// version so existing clients can continue to work without reconfiguration.
    /// </summary>
    /// <param name="oldName">Current collection name.</param>
    /// <param name="newName">New collection name.</param>
    /// <param name="cancellationToken">Optional cancellation token.</param>
    public async Task RenameCollectionAsync(
        string oldName,
        string newName,
        CancellationToken cancellationToken = default)
    {
        var body = new Dictionary<string, string> { ["new_name"] = newName };
        await RequestAsync<object>(
            "POST",
            $"/collections/{Uri.EscapeDataString(oldName)}/rename",
            body,
            cancellationToken);
    }

    /// <summary>
    /// Rebuilds the HNSW index for a collection with new parameters
    /// (POST /collections/{name}/reindex).
    /// No re-embedding is required — existing stored vectors are reused.
    /// Returns a <see cref="ReindexJob"/> with State == "completed" on success.
    /// </summary>
    /// <param name="name">Collection name.</param>
    /// <param name="parameters">HNSW rebuild parameters.</param>
    /// <param name="cancellationToken">Optional cancellation token.</param>
    public async Task<ReindexJob> ReindexCollectionAsync(
        string name,
        ReindexParams parameters,
        CancellationToken cancellationToken = default)
    {
        return await RequestAsync<ReindexJob>(
            "POST",
            $"/collections/{Uri.EscapeDataString(name)}/reindex",
            parameters,
            cancellationToken);
    }

    /// <summary>
    /// Creates a native per-collection snapshot (POST /collections/{name}/snapshot).
    /// The server writes a gzip-compressed JSON snapshot and returns snapshot
    /// metadata including ID, collection name, creation time, and size.
    /// </summary>
    /// <param name="name">Collection name.</param>
    /// <param name="cancellationToken">Optional cancellation token.</param>
    public async Task<NativeSnapshotInfo> SnapshotCollectionNativeAsync(
        string name,
        CancellationToken cancellationToken = default)
    {
        return await RequestAsync<NativeSnapshotInfo>(
            "POST",
            $"/collections/{Uri.EscapeDataString(name)}/snapshot",
            new Dictionary<string, object>(),
            cancellationToken);
    }

    /// <summary>
    /// Lists all native snapshots for a collection (GET /collections/{name}/snapshots).
    /// Returns snapshots newest-first as reported by the server.
    /// </summary>
    /// <param name="name">Collection name.</param>
    /// <param name="cancellationToken">Optional cancellation token.</param>
    public async Task<IReadOnlyList<NativeSnapshotInfo>> ListCollectionSnapshotsNativeAsync(
        string name,
        CancellationToken cancellationToken = default)
    {
        var envelope = await RequestAsync<SnapshotsEnvelope>(
            "GET",
            $"/collections/{Uri.EscapeDataString(name)}/snapshots",
            null,
            cancellationToken);
        return envelope.Snapshots ?? new List<NativeSnapshotInfo>();
    }

    /// <summary>
    /// Restores a collection from a native snapshot
    /// (POST /collections/{name}/snapshots/{id}/restore).
    /// Drops the current in-memory state and replaces it with the snapshot data.
    /// </summary>
    /// <param name="name">Collection name.</param>
    /// <param name="snapshotId">Snapshot identifier returned by <see cref="SnapshotCollectionNativeAsync"/>.</param>
    /// <param name="cancellationToken">Optional cancellation token.</param>
    public async Task RestoreCollectionSnapshotNativeAsync(
        string name,
        string snapshotId,
        CancellationToken cancellationToken = default)
    {
        await RequestAsync<object>(
            "POST",
            $"/collections/{Uri.EscapeDataString(name)}/snapshots/{Uri.EscapeDataString(snapshotId)}/restore",
            new Dictionary<string, object>(),
            cancellationToken);
    }

    /// <summary>
    /// Runs a vector search and returns the full HNSW execution trace
    /// (POST /collections/{name}/explain).
    /// The trace includes visited_nodes, ef_search, hnsw_search_ms,
    /// payload_filter_evals, quantization_score_ms, and total_ms.
    /// </summary>
    /// <param name="collection">Collection name.</param>
    /// <param name="vector">Query vector.</param>
    /// <param name="k">Number of nearest neighbours to retrieve.</param>
    /// <param name="cancellationToken">Optional cancellation token.</param>
    public async Task<ExplainResponse> ExplainSearchAsync(
        string collection,
        IReadOnlyList<float> vector,
        int k,
        CancellationToken cancellationToken = default)
    {
        var body = new Dictionary<string, object>
        {
            ["vector"] = vector,
            ["k"] = k
        };
        return await RequestAsync<ExplainResponse>(
            "POST",
            $"/collections/{Uri.EscapeDataString(collection)}/explain",
            body,
            cancellationToken);
    }

    /// <summary>
    /// Returns entries from the slow-query ring buffer (GET /slow_queries).
    /// Entries are returned in the order they were recorded (oldest first).
    /// Use <see cref="SetSlowQueryConfigAsync"/> to tune the threshold and capacity.
    /// </summary>
    /// <param name="cancellationToken">Optional cancellation token.</param>
    public async Task<IReadOnlyList<SlowQueryEntry>> ListSlowQueriesAsync(
        CancellationToken cancellationToken = default)
    {
        var envelope = await RequestAsync<SlowQueryEntriesEnvelope>(
            "GET",
            "/slow_queries",
            null,
            cancellationToken);
        return envelope.Entries ?? new List<SlowQueryEntry>();
    }

    /// <summary>
    /// Reconfigures the slow-query ring buffer (POST /slow_queries/config).
    /// Existing entries are retained. If the new capacity is smaller than the
    /// current entry count the oldest entries are evicted by the server.
    /// Returns the updated configuration as echoed back by the server.
    /// </summary>
    /// <param name="config">New slow-query configuration.</param>
    /// <param name="cancellationToken">Optional cancellation token.</param>
    public async Task<SlowQueryConfig> SetSlowQueryConfigAsync(
        SlowQueryConfig config,
        CancellationToken cancellationToken = default)
    {
        return await RequestAsync<SlowQueryConfig>(
            "POST",
            "/slow_queries/config",
            config,
            cancellationToken);
    }
}
